astropy/timeseries/tests/test_downsample.py

```diff
@@ -33,38 +39,123 @@ def test_timeseries_invalid():
+
+
+def test_binning_arg_invalid():
+    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5]], names=['a'])
+    with pytest.raises(TypeError, match=r"With single 'time_bin_start' either 'n_bins', "
+                                         "'time_bin_size' or time_bin_end' must be provided"):
+        aggregate_downsample(ts)
+
+
+def test_time_bin_conversion():
+
+    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5]], names=['a'])
+
+    # Make sure time_bin_start and time_bin_end are properly converted to Time
+    down_start = aggregate_downsample(ts, time_bin_start=['2016-03-22T12:30:31'],
+                                      time_bin_size=[1]*u.s)
+    assert_equal(down_start.time_bin_start.isot, ['2016-03-22T12:30:31.000'])
+
+    down_end = aggregate_downsample(ts, time_bin_start=['2016-03-22T12:30:31', '2016-03-22T12:30:33'],
+                                    time_bin_end='2016-03-22T12:30:34')
+    assert_equal(down_end.time_bin_end.isot, ['2016-03-22T12:30:33.000', '2016-03-22T12:30:34.000'])
+
+
+def test_time_bin_end_auto():
+
+    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5]], names=['a'])
+
+    # Interpret `time_bin_end` as the end of timeseries when `time_bin_start` is
+    # an array and `time_bin_size` is not provided
+    down_auto_end = aggregate_downsample(ts, time_bin_start=['2016-03-22T12:30:31', '2016-03-22T12:30:33'])
+    assert_equal(down_auto_end.time_bin_end.isot, ['2016-03-22T12:30:33.000', '2016-03-22T12:30:35.000'])
+
+
+def test_time_bin_start_array():
+
+    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5]], names=['a'])
+
+    # When `time_bin_end` is an array and `time_bin_start` is not provided, `time_bin_start` is converted
+    # to an array with its first element set to the start of the timeseries and rest populated using
+    #`time_bin_end`. This case is separately tested since `BinnedTimeSeries` allows `time_bin_end` to
+    # be an array only if `time_bin_start` is an array.
+    down_start_array = aggregate_downsample(ts, time_bin_end=['2016-03-22T12:30:33', '2016-03-22T12:30:35'])
+    assert_equal(down_start_array.time_bin_start.isot, ['2016-03-22T12:30:31.000', '2016-03-22T12:30:33.000'])
+
+
+def test_nbins():
+
+    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5]], names=['a'])
+
+    # n_bins should default to the number needed to fit all the original points
+    down_nbins = aggregate_downsample(ts, n_bins=2)
+    assert_equal(down_nbins.time_bin_start.isot, ['2016-03-22T12:30:31.000', '2016-03-22T12:30:33.000'])
 
 
 def test_downsample():
-    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4]], names=['a'])
-    ts_units = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4] * u.count], names=['a'])
+    ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5]], names=['a'])
+    ts_units = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4, 5] * u.count], names=['a'])
 
     down_1 = aggregate_downsample(ts, time_bin_size=1*u.second)
```

Comment:
```suggestion
    if np.finfo(float).dtype == np.float32:
        time_bin_incr = (1 - 1e-6) * u.s
    else:
        time_bin_incr = 1 * u.s

    down_1 = aggregate_downsample(ts, time_bin_size=1*time_bin_incr)
```
Sorry if I haven’t been clear; my computer has all but blown up and it’s difficult to see on a phone what exactly your latest changes were or edit in the GH interface. But what I meant was, don’t change `INPUT_TIME` at all but instead set time_bin_size this way, and then with multiples 2, 3 etc.
If that doesn’t work, yes, skipping is ok with me.

Reply:
Aah, that makes sense. I thought you were suggesting changes the `INPUT TIMES`.

Reply:
That still doesn't work 😞 . I think I will skip the test.